archive/100-theorems-list/45_partition.lean

```diff
@@ -0,0 +1,516 @@
+               nat.mem_antidiagonal, prod.exists],
+    refine ⟨f a, s.sum f, rfl, λ i, if i = a then 0 else f i, _, _⟩,
+    { rw [mem_cut],
+      refine ⟨_, _⟩,
+      { rw [sum_ite],
+        have : (filter (λ x, x ≠ a) s) = s,
+        { apply filter_true_of_mem,
+          rintro i hi rfl,
+          apply h hi },
+        simp [this] },
+      { intros i hi,
+        rw ite_eq_left_iff,
+        intro hne,
+        apply h₁,
+        simp [not_or_distrib, hne, hi] } },
+    { ext,
+      obtain rfl|h := eq_or_ne x a,
+      { simp },
+      { simp [if_neg h] } } },
+  { simp only [mem_insert, function.embedding.coe_fn_mk, mem_map, nat.mem_antidiagonal, prod.exists,
+               exists_prop, mem_cut, not_or_distrib],
+    rintro ⟨p, q, rfl, g, ⟨rfl, hg₂⟩, rfl⟩,
+    refine ⟨_, _⟩,
+    { simp [sum_add_distrib, if_neg h, hg₂ _ h, add_comm] },
+    { rintro i ⟨h₁, h₂⟩,
+      simp [if_neg h₁, hg₂ _ h₂] } }
+end
+
+lemma coeff_prod_range
+  [comm_semiring α] {ι : Type*} (s : finset ι) (f : ι → power_series α) (n : ℕ) :
+  coeff α n (∏ j in s, f j) = ∑ l in cut s n, ∏ i in s, coeff α (l i) (f i) :=
+begin
+  revert n,
+  apply finset.induction_on s,
+  { rintro ⟨_ | n⟩,
+    { simp },
+    simp [cut_empty_succ, if_neg (nat.succ_ne_zero _)] },
+  intros a s hi ih n,
+  rw [cut_insert _ _ _ hi, prod_insert hi, coeff_mul, sum_bUnion],
+  { apply sum_congr rfl _,
+    simp only [prod.forall, sum_map, pi.add_apply,
+               function.embedding.coe_fn_mk, nat.mem_antidiagonal],
+    rintro i j rfl,
+    simp only [prod_insert hi, if_pos rfl, ih, mul_sum],
+    apply sum_congr rfl _,
+    intros x hx,
+    rw mem_cut at hx,
+    rw [hx.2 a hi, zero_add],
+    congr' 1,
+    apply prod_congr rfl,
+    intros k hk,
+    rw [if_neg, add_zero],
+    exact ne_of_mem_of_not_mem hk hi },
+  { simp only [set.pairwise_disjoint, set.pairwise, prod.forall, not_and, ne.def,
+      nat.mem_antidiagonal, disjoint_left, mem_map, exists_prop, function.embedding.coe_fn_mk,
+      exists_imp_distrib, not_exists, finset.mem_coe],
+    rintro p₁ q₁ rfl p₂ q₂ h t x hx,
+    simp only [finset.inf_eq_inter, finset.mem_map, finset.mem_inter, mem_cut, exists_prop,
+      function.embedding.coe_fn_mk] at hx,
+    rcases hx with ⟨⟨p, ⟨hp, hp2⟩, hp3⟩, ⟨q, ⟨hq, hq2⟩, hq3⟩⟩,
+    have z := hp3.trans hq3.symm,
+    have := sum_congr (eq.refl s) (λ x _, function.funext_iff.1 z x),
+    obtain rfl : q₁ = q₂,
+    { simpa [sum_add_distrib, hp, hq, if_neg hi] using this },
+    obtain rfl : p₂ = p₁,
+    { simpa using h },
+    exact (t rfl).elim }
+end
+
+/-- A convience constructor for the power series whose coefficients indicate a subset. -/
+def indicator_series (α : Type*) [semiring α] (f : set ℕ) : power_series α :=
+power_series.mk (λ n, if f n then 1 else 0)
+
+lemma coeff_indicator (f : set ℕ) [semiring α] (n : ℕ) :
+  coeff α n (indicator_series _ f) = if f n then 1 else 0 :=
+coeff_mk _ _
+lemma coeff_indicator_pos (f : set ℕ) [semiring α] (n : ℕ) (h : f n):
```

Comment:
```suggestion
lemma coeff_indicator_pos (f : set ℕ) [semiring α] (n : ℕ) (h : f n):
```
Please distinguish between `set ℕ` and `ℕ -> Prop`. Since `f : set ℕ`, write `n \in f` (maybe change the letter `f` so it looks like a set, not a function).
This applies to many places

Reply:
Good point, I think I changed them all now